File: electron-wrapper/proxy/singbox-supervisor.js
'use strict';

const fs = require('fs');
const path = require('path');
const child_process = require('child_process');

const SOCKS_PORT = 1080;
const HTTP_PORT = 1081;
const MAX_RESTARTS_PER_MINUTE = 3;

/**
 * Owns the sing-box subprocess. Writes config JSON, spawns the binary, and
 * restarts on crash. Mirrors the Dart-side SingboxSupervisor (Task 16):
 * 1 initial start + 3 restarts in a rolling 60s window before giving up.
 */
class SingboxSupervisor {
  constructor({ binaryPath, configDir, spawnFn }) {
    this.binaryPath = binaryPath;
    this.configDir = configDir;
    this._spawn = spawnFn || child_process.spawn;
    this._currentNode = null;
    this._proc = null;
    this._recentStarts = [];
    this._stopped = false;
    this.socksAddress = `127.0.0.1:${SOCKS_PORT}`;
    this.httpAddress = `127.0.0.1:${HTTP_PORT}`;
    this._gaveUpResolve = null;
    this._gaveUpReject = null;
    this.gaveUp = new Promise((resolve, reject) => {
      this._gaveUpResolve = resolve;
      this._gaveUpReject = reject;
    });
  }

  async startWith(node) {
    this._currentNode = node;
    await this._spawnOne();
  }

  async _spawnOne() {
    if (this._stopped) return;
    if (this._proc) {
      try { this._proc.removeAllListeners('exit'); } catch (_) { /* swallow */ }
      try { this._proc.kill(); } catch (_) { /* swallow */ }
      this._proc = null;
    }
    const cfgPath = path.join(this.configDir, 'singbox-config.json');
    fs.writeFileSync(cfgPath, this._renderConfig(this._currentNode), { mode: 0o600 });
    if (process.platform !== 'win32') {
      try { fs.chmodSync(cfgPath, 0o600); } catch (_) { /* best-effort */ }
    }
    this._recentStarts.push(Date.now());

    const node = this._currentNode || {};
    console.log('[sing-box] spawn binary=%s cfg=%s node=%s server=%s:%d cipher=%s',
      this.binaryPath, cfgPath, node.name, node.server, node.port, node.cipher);

    const proc = this._spawn(this.binaryPath, ['run', '-c', cfgPath]);
    this._proc = proc;
    if (proc && proc.stdout && typeof proc.stdout.on === 'function') {
      proc.stdout.on('data', (chunk) => {
        process.stdout.write('[sing-box stdout] ' + chunk.toString());
      });
    }
    if (proc && proc.stderr && typeof proc.stderr.on === 'function') {
      proc.stderr.on('data', (chunk) => {
        process.stderr.write('[sing-box stderr] ' + chunk.toString());
      });
    }
    if (proc && typeof proc.on === 'function') {
      proc.on('error', (err) => {
        console.error('[sing-box] spawn error:', err);
      });
    }
    proc.on('exit', (code, signal) => {
      console.warn('[sing-box] exit code=%s signal=%s stopped=%s', code, signal, this._stopped);
      if (this._stopped) return;
      this._onExit();
    });
  }

  _onExit() {
    const now = Date.now();
    this._recentStarts = this._recentStarts.filter(t => now - t < 60_000);
    console.warn('[sing-box] _onExit; restarts in last 60s=%d (limit=%d)',
      this._recentStarts.length, MAX_RESTARTS_PER_MINUTE + 1);
    if (this._recentStarts.length >= MAX_RESTARTS_PER_MINUTE + 1) {
      console.error('[sing-box] restart budget exhausted → gaveUp');
      if (this._gaveUpResolve) {
        this._gaveUpResolve();
        this._gaveUpResolve = null;
        this._gaveUpReject = null;
      }
      return;
    }
    console.log('[sing-box] respawning…');
    this._spawnOne().catch(err => {
      console.error('[sing-box] respawn failed:', err);
      if (this._gaveUpReject) {
        this._gaveUpReject(err);
        this._gaveUpResolve = null;
        this._gaveUpReject = null;
      }
    });
  }

  _renderConfig(n) {
    return JSON.stringify({
      log: { level: 'info', disabled: false, timestamp: true },
      inbounds: [
        { type: 'socks', tag: 'socks-in', listen: '127.0.0.1', listen_port: SOCKS_PORT },
        { type: 'http', tag: 'http-in', listen: '127.0.0.1', listen_port: HTTP_PORT },
      ],
      outbounds: [
        {
          type: 'shadowsocks',
          tag: 'proxy-out',
          server: n.server,
          server_port: n.port,
          method: n.cipher,
          password: n.password,
        },
      ],
    }, null, 2);
  }

  async stop() {
    this._stopped = true;
    if (this._proc) {
      try { this._proc.removeAllListeners('exit'); } catch (_) { /* swallow */ }
      try { this._proc.kill(); } catch (_) { /* swallow */ }
      this._proc = null;
    }
    // Resolve gaveUp so any .then() handlers attached via attachGaveUp run
    // once and release their closures. Otherwise the Promise stays pending
    // forever and across retry() cycles each ProxyRuntime accumulates one
    // dangling handler per discarded supervisor (slow leak).
    if (this._gaveUpResolve) {
      this._gaveUpResolve();
      this._gaveUpResolve = null;
      this._gaveUpReject = null;
    }
  }
}

module.exports = { SingboxSupervisor };

File: flutter-wrapper/lib/proxy/singbox_supervisor.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'file_logger.dart';
import 'proxy_node.dart';

/// Factory for creating a process. Returns a handle that exposes
/// `int get pid`, `Future<int> get exitCode`, and `void kill()`.
///
/// In production this is backed by `Process.start` (see [_RealProcessAdapter]).
/// Tests inject a fake process whose interface is structurally compatible —
/// hence `dynamic` here, so the fake doesn't need to extend a Dart class.
typedef ProcessFactory = Future<dynamic> Function(
  String binary,
  List<String> args,
);

class _RealProcessAdapter {
  final Process _proc;
  _RealProcessAdapter(this._proc) {
    // Tee sing-box's own stdout/stderr (it logs DNS lookups, dial errors,
    // outbound failures, etc.) into our durable file logger so the user can
    // diagnose without an attached debugger. sing-box already prints with its
    // own timestamps; we add level=info/error and a tag so the merged log is
    // greppable.
    _proc.stdout.transform(utf8.decoder).transform(const LineSplitter()).listen(
      (line) => pLog('info', '[sing-box] $line'),
      onError: (Object e) => pLog('error', '[sing-box] stdout error: $e'),
    );
    _proc.stderr.transform(utf8.decoder).transform(const LineSplitter()).listen(
      (line) => pLog('warn', '[sing-box] $line'),
      onError: (Object e) => pLog('error', '[sing-box] stderr error: $e'),
    );
  }
  int get pid => _proc.pid;
  Future<int> get exitCode => _proc.exitCode;
  void kill() => _proc.kill();
}

/// Supervises a sing-box process: writes its config, spawns it, and
/// restarts on crash up to [_maxRestartsPerMinute] times within a 60s
/// rolling window. Once that limit is exceeded, [gaveUp] resolves and
/// no further restarts happen.
class SingboxSupervisor {
  static const _socksPort = 1080;
  static const _httpPort = 1081;
  static const _maxRestartsPerMinute = 3;

  final String binaryPath;
  final String configDir;
  final ProcessFactory _factory;

  ProxyNode? _currentNode;
  dynamic _process;
  StreamSubscription<int>? _exitSub;
  final List<DateTime> _recentStarts = [];
  final Completer<void> _gaveUp = Completer<void>();
  bool _stopped = false;

  SingboxSupervisor({
    required this.binaryPath,
    required this.configDir,
    ProcessFactory? processFactory,
  }) : _factory = processFactory ?? _spawnReal;

  static Future<dynamic> _spawnReal(String binary, List<String> args) async {
    final p = await Process.start(binary, args);
    return _RealProcessAdapter(p);
  }

  Future<void> get gaveUp => _gaveUp.future;
  String get socksAddress => '127.0.0.1:$_socksPort';
  String get httpAddress => '127.0.0.1:$_httpPort';

  Future<void> startWith(ProxyNode node) async {
    _currentNode = node;
    await _spawn();
  }

  Future<void> _spawn() async {
    if (_stopped) return;
    final node = _currentNode;
    if (node != null) {
      pLogf(
        'info',
        '[supervisor] spawn binary=%s socks=%s http=%s node=%s server=%s:%s cipher=%s',
        [binaryPath, socksAddress, httpAddress, node.name, node.server, node.port, node.cipher],
      );
    }
    // If a previous process is still running (e.g. from probe-selector
    // walking through candidate nodes), terminate it before spawning the
    // next one. Otherwise we accumulate orphan sing-box processes.
    // Cancellation is fire-and-forget: under the test fakes the source
    // Future never completes, so awaiting cancel() can hang. The _stopped
    // flag plus nulling the ref are what actually prevent stale callbacks.
    _exitSub?.cancel();
    _exitSub = null;
    try {
      (_process as dynamic)?.kill();
    } catch (_) {}

    final cfgPath = '$configDir/singbox-config.json';
    await File(cfgPath).writeAsString(_renderConfig(_currentNode!));
    // Restrict permissions: the config contains the proxy password.
    // chmod is a no-op on Windows; the runProcess returns silently if it fails,
    // which is acceptable since the leaked-permission consequence is the same.
    if (!Platform.isWindows) {
      try {
        final result = await Process.run('chmod', ['600', cfgPath]);
        // Surface non-zero exit in debug/test builds so a silent chmod
        // failure (e.g. chmod absent from PATH on some Android variants)
        // doesn't leave the password-bearing config world-readable
        // without any signal.
        assert(
          result.exitCode == 0,
          'chmod 600 failed (exit ${result.exitCode}): ${result.stderr}',
        );
      } catch (_) { /* best-effort */ }
    }
    _recentStarts.add(DateTime.now());

    final handle = await _factory(binaryPath, ['run', '-c', cfgPath]);
    _process = handle;

    final exitFuture = handle.exitCode as Future<int>;
    _exitSub = exitFuture.asStream().listen((_) {
      if (_stopped) return;
      _onExit();
    });
  }

  void _onExit() {
    final now = DateTime.now();
    _recentStarts.removeWhere(
      (t) => now.difference(t) > const Duration(minutes: 1),
    );
    // 1 initial start + _maxRestartsPerMinute restarts = threshold.
    // If we've started more than that in the last minute, give up.
    if (_recentStarts.length >= _maxRestartsPerMinute + 1) {
      if (!_gaveUp.isCompleted) _gaveUp.complete();
      return;
    }
    // Fire-and-forget restart, but route any exception (file write failure,
    // factory throw) into gaveUp so callers waiting on it don't hang.
    _spawn().catchError((Object e, StackTrace st) {
      if (!_gaveUp.isCompleted) _gaveUp.completeError(e, st);
    });
  }

  String _renderConfig(ProxyNode node) {
    return const JsonEncoder.withIndent('  ').convert({
      // 'info' so dial errors / dns lookups for the SS server show up in our
      // tee'd log; the file rotates at 1MB so noise is bounded.
      'log': {'level': 'info', 'disabled': false, 'timestamp': true},
      'inbounds': [
        {
          'type': 'socks',
          'tag': 'socks-in',
          'listen': '127.0.0.1',
          'listen_port': _socksPort,
        },
        {
          'type': 'http',
          'tag': 'http-in',
          'listen': '127.0.0.1',
          'listen_port': _httpPort,
        },
      ],
      'outbounds': [
        {
          'type': 'shadowsocks',
          'tag': 'proxy-out',
          'server': node.server,
          'server_port': node.port,
          'method': node.cipher,
          'password': node.password,
        }
      ],
    });
  }

  Future<void> stop() async {
    _stopped = true;
    _exitSub?.cancel();
    _exitSub = null;
    try {
      (_process as dynamic)?.kill();
    } catch (_) {}
  }
}
